dan_layer/engine_types/src/substate.rs

```diff
@@ -351,33 +351,45 @@ impl FromStr for SubstateId {
                 Ok(SubstateId::Component(addr))
             },
             Some(("resource", addr)) => {
-                match addr.split_once(' ') {
-                    Some((resource_str, addr)) => match addr.split_once('_') {
-                        // resource_xxxx nft_xxxxx
-                        Some(("nft", _)) => {
-                            let nft_address =
-                                NonFungibleAddress::from_str(s).map_err(|e| InvalidSubstateIdFormat(e.to_string()))?;
-                            Ok(SubstateId::NonFungible(nft_address))
-                        },
-                        // resource_xxxx index_
-                        Some(("index", index_str)) => {
-                            let resource_addr = ResourceAddress::from_hex(resource_str)
-                                .map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
-                            let index = u64::from_str(index_str).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
-                            Ok(SubstateId::NonFungibleIndex(NonFungibleIndexAddress::new(
-                                resource_addr,
-                                index,
-                            )))
-                        },
-                        _ => Err(InvalidSubstateIdFormat(s.to_string())),
-                    },
-                    // resource_xxxx
-                    None => {
-                        let addr =
-                            ResourceAddress::from_hex(addr).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
-                        Ok(SubstateId::Resource(addr))
-                    },
+                if addr.split_once('_').is_none() {
+                    // resource_xxxxx
+                    let addr = ResourceAddress::from_hex(addr).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
+                    return Ok(SubstateId::Resource(addr));
                 }
+                Err(InvalidSubstateIdFormat("Missing resource address".to_string()))
```

Comment:
from_hex will already error if addr is malformed. The error "Resource address is missing" doesn't really tell us much. InvalidSubstateIdFormat error type is meant to have the string that we tried to parse in it. e.g. will display `Invalid substate ID "resource_deadbeaf_bad"`. 

```suggestion
                    // resource_xxxxx
                    let addr = ResourceAddress::from_hex(addr).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
                    return Ok(SubstateId::Resource(addr));
```